File: Design-Patterns-Using-Java/Basics OOPS/Inheritance.java
class Person {
    protected String name;
    protected int age;
    protected String gender;

    public Person(String name, int age, String gender) {
        this.name = name;
        this.age = age;
        this.gender = gender;
    }

    public String toString() {
        return String.format("Person: %s is a %s and %s years old.", name, gender, age);
    }

    public String getGender() {
        return gender;
    }
}

class Citizen extends Person {
    private int id;

    public Citizen(String name, int age, String gender, int id) {
        super(name, age, gender);
        this.id = id;
    }

    public int getVoterId() {
        return id;
    }

    @Override
    public String toString() {
        return String.format("Citizen: %s is a %s and %s years old with voter id %s.", name, gender, age, id);
    }
}

public class Inheritance {
    public static void main(String[] args) {
        Person p = new Person("John", 32, "Male");
        System.out.println(p);

        Citizen c = new Citizen("Smith", 31, "Male", 1234);
        System.out.println(c);
    }
}

/* 
Person: John is a Male and 32 years old.
Citizen: Smith is a Male and 31 years old with voter id 1234. 
*/

File: Design-Patterns-Using-CSharp/Basics OOPS/Inheritance.cs
using System;

class Person
{
    protected string name;
    protected int age;
    protected string gender;

    public Person(string name, int age, string gender)
    {
        this.name = name;
        this.age = age;
        this.gender = gender;
    }

    public override string ToString()
    {
        return string.Format("Person: {0} is a {1} and {2} years old.", name, gender, age);
    }

    public string GetGender()
    {
        return gender;
    }
}

class Citizen : Person
{
    private int id;

    public Citizen(string name, int age, int id, string gender) : base(name, age, gender)
    {
        this.id = id;
    }

    public int GetVoterId()
    {
        return id;
    }

    public override string ToString()
    {
        return string.Format("Citizen: {0} is a {1} and {2} years old with voter id {3}.", name, gender, age, id);
    }
}

public class Inheritance
{
    public static void Main(string[] args)
    {
        Person p = new Person("John", 32, "Male");
        Console.WriteLine(p);

        Citizen c = new Citizen("Smith", 31, 1234, "Male");
        Console.WriteLine(c);
    }
}
